Add file system view desktop user interface

ianyong opened the pull request:
### Description

Add the file system view for the desktop user interface. Similar to most code editors/IDEs available today, the file system view can be found within a sidebar on the left edge of the screen. This sidebar is built to be generic so that future features such as git integration can also make use of it.

*Note: the sidebar tab is now named `Files` instead of `Project` after being deemed to be confusing but the below screenshots are not updated since everything else is still the same.*

![image](https://user-images.githubusercontent.com/5585517/194744795-e010a77f-760e-4dad-bd26-0b7904305e23.png)
![vmconnect_9M3ear1SQ1](https://user-images.githubusercontent.com/5585517/194745228-ffbee7ff-8ec6-4bb8-b47e-e3d29f029605.gif)

Some of the move complicated design decisions were documented in the wiki, namely the [handling of file system updates](https://github.com/source-academy/frontend/wiki/File-System#handling-file-system-updates) and the [lack of options in the BrowserFS API](https://github.com/source-academy/frontend/wiki/File-System#lack-of-options).

This PR adds a new dependency on [react-menu](https://github.com/szhsin/react-menu) due to it being a highly customisable React library for handling context menus (i.e., right click menus) which are used a lot in the file system view. @martin-henz has previously approved the use of this dependency.

Part of #2176.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [ ] Code quality improvements

### How to test

Open the playground and play around with the user interface. Note that the editor is not linked to the file system yet and will not change its contents based on interactions with the file system view.

### Changes that will be tackled in a future PR

* Mobile user interface.
* Resizing of sidebar.
  * Blocked by #2234.
* Different icons based on file extension.
  * Very low priority.
* Hooking up the file system view with the code editor.
* Removing the flickering whenever a file system operation is carried out. This is because the BrowserFS operations are asynchronous.
  * Low priority.

ianyong commented:
@chownces @zhaojj2209 If you have any feedback on the UI/UX (not just the code), do let me know too!

zhaojj2209 reviewed (commented):
Code wise mostly looks good, UI wise some suggestions:
- Hover does not extend to the end of the component (replicable on Chrome and Safari). I suspect it's because of the vertical scrollbar.
<img width="618" alt="Screenshot 2022-10-09 at 23 14 34" src="https://user-images.githubusercontent.com/54243224/194765302-587ff54b-c79f-4756-8ed8-3ea217ed4a69.png">

- Would be good to disable spell-check since it's likely the red line will pop up for every file with file extensions. Red line only shows up on Chrome and Safari.

https://user-images.githubusercontent.com/54243224/194765329-3b457c5c-2c45-4439-86c8-74444388a9c3.mov

- Folders are currently togglable while renaming, could be a minor annoyance.

https://user-images.githubusercontent.com/54243224/194765416-5b4a26f9-37b1-4a16-b00d-28b520288f0a.mov

- Folder system does not seem to work on MS Edge on Mac. Not sure if this is important to fix, though, given that Mac users are unlikely to use MS Edge.

![photo_2022-10-09 23 31 05](https://user-images.githubusercontent.com/54243224/194765469-f669ceb6-f0bb-4624-8a23-300ab126bd84.jpeg)

- I think it would be good to add buttons for creating files/directories at the top of the file system view, similar to that in VSCode. This can be a future improvement added in a separate PR.
<img width="60" alt="Screenshot 2022-10-09 at 23 13 45" src="https://user-images.githubusercontent.com/54243224/194765689-b19340c6-1b3a-41e9-81b2-7b0206c00f65.png">

chownces reviewed (approved):
LGTM! I will do a full QA test on existing features when we are merging into master. Shall we take a look into feature toggles to disable these features before they are ready?

ianyong commented on src/commons/delay/Delay.tsx:
```diff
@@ -0,0 +1,21 @@
+  waitBeforeRender: number;
+};
+
+const Delay: React.FC<DelayProps> = (props: DelayProps) => {
```
I made the unit of time more obvious by renaming `waitBeforeRender` to `waitInMsBeforeRender`, as well as added the following header comment:
```ts
/**
 * Delays the rendering of child components by a set time.
 */
```

I'm not very sure if this is what you expect since the component is very simple and its usage should be easily understood. If what you mean is to touch on the concept of composition (i.e., https://reactjs.org/docs/composition-vs-inheritance.html), I don't think this is the right place to do so because composition is a general concept that is widely used across React applications.

zhaojj2209 replied:
Comment looks good, thanks for the changes!

ianyong commented:
Thanks for testing on so many different browsers! :0

> * Hover does not extend to the end of the component (replicable on Chrome and Safari). I suspect it's because of the vertical scrollbar.

It seems that unlike on Firefox, the scrollbars in Chromium browsers (and presumably Safari which I cannot test on) always take up space which results in this behaviour. I've changed the `overflow` CSS property from `scroll` to `auto` so that hidden scrollbars do not take up space. However, scrollbars that do exist still take up space; see the following screenshot:

![image](https://user-images.githubusercontent.com/5585517/194808241-5cb92a22-cef9-4fb3-8349-aa75a4bf6cfe.png)

Separately, there seems to be a white square at the intersection of the x-axis and y-axis scrollbars on Chromium browsers. I've added a CSS rule within the context of the file system view to make the background of that square transparent.

> * Would be good to disable spell-check since it's likely the red line will pop up for every file with file extensions. Red line only shows up on Chrome and Safari.

Nice catch! I've disabled spellcheck in the `Input` elements.

> * Folders are currently togglable while renaming, could be a minor annoyance.

Agreed, I've disabled toggling when renaming.

> * Folder system does not seem to work on MS Edge on Mac. Not sure if this is important to fix, though, given that Mac users are unlikely to use MS Edge.

I could get it to work on Microsoft Edge Version 102.0.1245.41 (Official build) (64-bit) running on Ubuntu. Are there any errors in the console. or any other information I can work off?

ianyong commented:
> Shall we take a look into feature toggles to disable these features before they are ready?

I don't think this would be a good use of time because we will eventually need to implement some kind of toggle that is either tied to the Source chapter or course progression. The merging into master despite the feature not being ready is so as to avoid a big bang integration at the very end which might have lots of merge conflicts. For now, the feature is not meant to be used at all in production and I will comment out the code responsible for integrating the feature into the rest of the Source Academy.

zhaojj2209 reviewed (approved):
LGTM, thanks for the changes!

> I could get it to work on Microsoft Edge Version 102.0.1245.41 (Official build) (64-bit) running on Ubuntu. Are there any errors in the console. or any other information I can work off?

No errors in console, seems that the BrowserFS `configure` function executed without throwing any errors yet the file system is `undefined`. Nobody uses MSEdge on Mac though, so as long as it works on Windows/Linux it shouldn't be an issue.
